iOS Compatibility

swoolcock opened the pull request:
Solution/projects/changes to run osu!lazer on iOS devices.

- [x] Depends on https://github.com/ppy/osu-framework/pull/1409

Tasks:
- [x] Login
- [x] Chat
- [x] osu!direct search and downloads
- [x] osu!standard gameplay
- [x] osu!taiko gameplay
- [x] Non-ruleset visual tests

- [x] Depends on #4147 
- [x] Depends on #4153 
- [x] Depends on #3919 


To be done post-merge:
- osu!mania gameplay (except hold notes)
- osu!catch gameplay (see below)
- Ruleset visual tests (see below)

Issues:
- ~~osu!standard spinners crash the app at the end of the spinner unless you don't spin it at all~~
- ~~Anything using `BufferedContainer`s, due to issues outlined in the osu!framework PR~~
- Currently can't implement osu!mania hold notes or osu!catch (need to do some hacks to listen for individual key up/down events, as iOS does not expose them)
- Ruleset visual tests not currently building
- Visual tests that require file system access may not run due to permissions or non-existent files

swoolcock committed:
Add osu.iOS project

swoolcock committed:
Make osu! run. Run osu!, run.

swoolcock committed:
Don’t directly instantiate TrackBass

swoolcock committed:
Merge branch 'master' into samah-ios

# Conflicts:
#	osu.Game/osu.Game.csproj

swoolcock committed:
Merge branch 'master' into samah-ios

# Conflicts:
#	osu.Game/osu.Game.csproj

swoolcock committed:
Fix package references

swoolcock committed:
Add AppIcon

swoolcock committed:
Merge branch 'master' into samah-ios

swoolcock committed:
Split solution

swoolcock committed:
Merge branch 'master' into samah-ios

MiraiSubject commented:
I have almost been able to compile this version for iOS. However, I've been running in issues with Sixlabors/Core (and Imagesharp) versions that are not available on NuGet. 

Specifically speaking these entries in `osu.iOS/packages.config`: 

`
<package id="SixLabors.Core" version="1.0.0-dev000094" targetFramework="xamarinios10" /><package id="SixLabors.ImageSharp" version="1.0.0-dev002298" targetFramework="xamarinios10"/>`

On my copy of the solution I have been able to make some progress by using [this](https://www.nuget.org/packages/SixLabors.ImageSharp/1.0.0-beta0005) ImageSharp version instead. I have still been unsuccessful in getting the Sixlabors/Core properly linked because of this dependency issue I am having. 

For context, I've been using [62c4bc9](https://github.com/ppy/osu-framework/tree/62c4bc951996be6a3ed49b08db9acbf3151cf95b) (master at the time of writing) of ppy/osu-framework for the build process. 

osu.Desktop builds and runs successfully so it's isolated to the iOS build.

naoey commented:
@MiraiSubject you need to add the myget feed to your nuget sources (https://www.myget.org/F/sixlabors/api/v3/index.json)

MiraiSubject commented:
Thanks! @naoey I didn't think about that. Got it running on the 2018 iPad Pro now. Going to be looking into contributing now that I got this build working.

swoolcock commented:
> In osu!standard the game does not do anything after the spinner finishes on the loading screen for a beatmap.

This is to do with hovering not being detected, and was fixed in https://github.com/ppy/osu-framework/pull/2039

> Sometimes the game thread locks up and other times you can still interact with the elements.

I noticed this recently, it's something context-related I think.  The draw thread runs fine, but `SwapBuffers` seems to do nothing.  I'll make a ticket for it in the framework.

swoolcock committed:
Add basic visual tests

swoolcock committed:
Add ruleset visual tests

swoolcock committed:
Merge branch 'master' into samah-ios

# Conflicts:
#	osu.Game/osu.Game.csproj

swoolcock committed:
Use TrackManager CreateWaveform to fix editor crashes

swoolcock committed:
Merge branch 'master' into samah-ios

swoolcock committed:
Update framework references

swoolcock committed:
Revert to old behaviour if no TrackManager available

swoolcock committed:
Merge branch 'master' into samah-ios

swoolcock committed:
Update framework

swoolcock committed:
Ensure beatmap textures are loaded via CreateTextureLoaderStore

swoolcock committed:
Update framework

swoolcock committed:
Fix TrackManager usage

swoolcock committed:
Convert to PackageReference

swoolcock committed:
Revert audio changes

swoolcock committed:
Fix missing references

Note that the NativeLibs package required by osu.Framework includes assemblies that will not link on Xamarin.iOS.  For now we explicitly reference the package and exclude all assets, and manually provide the BASS libraries.

swoolcock committed:
Merge branch 'master' into samah-ios

# Conflicts:
#	osu.Game/osu.Game.csproj

swoolcock committed:
Update references

swoolcock committed:
Add better app icons and change info.plist properties

swoolcock committed:
Merge branch 'master' into samah-ios

swoolcock committed:
Update framework

swoolcock committed:
Update framework

swoolcock committed:
Code sanity

swoolcock committed:
Merge branch 'master' into samah-ios

swoolcock committed:
Pass GameHost to BeatmapManager in visual tests

swoolcock committed:
Merge branch 'master' into samah-ios

peppy committed:
Merge remote-tracking branch 'origin/master' into samah-ios

peppy committed:
Merge branch 'master' into samah-ios

peppy committed:
Merge remote-tracking branch 'origin/fix-idle-weirdness' into samah-ios

peppy committed:
Merge branch 'update-framework' into samah-ios

peppy committed:
Update framework version

peppy committed:
Update licence headers

peppy committed:
Merge remote-tracking branch 'origin/fix-texture-loader-stores' into samah-ios

# Conflicts:
#	osu.Game/Beatmaps/BeatmapManager.cs

peppy committed:
Merge branch 'dont-always-focus-textboxes' into samah-ios

swoolcock committed:
Move iOS package references to shared props

swoolcock committed:
Merge branch 'master' into samah-ios

peppy committed:
Merge branch 'master' into samah-ios

peppy committed:
Update resources

peppy committed:
Target iOS 10

peppy commented on osu.Game.Tests.iOS/osu.Game.Tests.iOS.csproj:
```diff
@@ -0,0 +1,87 @@
+    <Compile Include="AppDelegate.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="..\osu.Game.Tests\**\Beatmaps\**\*.cs">
```
is this because of some incompatible tests?

swoolcock replied:
For the unit tests to be discovered, they need to be in the executing assembly, it seems.  This means I couldn't split the tests into common, desktop and iOS projects.  Instead, I made an iOS project that directly links to the source and compiles with the same assembly name.

peppy replied:
I think it's probably best to not do this, as it's 100% going to get forgotten when we add a new folder or namespace.

swoolcock replied:
I wasn't particularly happy with it either.  Another option I thought of is to include `*.cs` across the entire project, then exclude anything it puts in the build directory.

peppy commented on osu.Game.Tests/Visual/TestCaseEditorComposeTimeline.cs:
```diff
@@ -30,7 +31,7 @@ public class TestCaseEditorComposeTimeline : EditorClockTestCase
 
         [BackgroundDependencyLoader]
-        private void load()
+        private void load(AudioManager audioManager)
```
not necessary right?

swoolcock replied:
Oops, must have left that in by mistake.

peppy committed:
Remove unnecessary changes

swoolcock committed:
Update iOS dotsettings file

peppy commented:
I've probably asked this already, but are `osu.iOS/libbass.a` still required locally?

swoolcock commented:
They are currently required, because of the explicit exclusion of `NativeLibs` in the iOS project.  Otherwise mtouch will try to link against libraries that are not supported for iOS.  We can potentially solve this by configuring `NativeLibs` to provide libraries based on target platform.

swoolcock committed:
Use project-wide *.cs linked include for tests rather than specifying subdirectories
